lib/datadog/appsec/waf.rb

```diff
@@ -349,7 +349,7 @@ def self.ruby_to_object(val, max_container_size: nil, max_container_depth: nil,
           obj
         when String
           obj = LibDDWAF::Object.new
-          val = val.to_s[0, max_string_length] if max_string_length
+          val = val.to_s.encode('utf-8', invalid: :replace, undef: :replace)[0, max_string_length] if max_string_length
```

Comment:
1. the conversion to utf-8 should be done regardless of the max_string_length, because the waf effectively expects utf-8
2. the slicing done on the string will operate on characters (unicode code points), whereas the waf limits are applied against the number of the bytes in the encoded string (which is possibly larger, up until 4x larger). So if you didn't change the default limits for the WAF, the string can possibly be truncated a second time inside the WAF. Personally, I just set very high limits on the WAF.

Reply:
> the conversion to utf-8 should be done regardless of the max_string_length

Good catch. @GustavoCaso's code above achieves that although that suggestion's initial goal was readability.

I will make the change.

> the slicing done on the string will operate on characters (unicode code points)

Not quite correct, a character may be made of multiple code points. We may end up with a sequence that has valid code points but that does not make sense in terms of characters.

> the waf limits are applied against the number of the bytes

Ah, so it wants UTF-8 but doesn't _always_ process it as such. I venture this consistency is being sacrificed on the altar of performance.

We use the internal defaults:

- https://github.com/DataDog/libddwaf-rb/blob/a518093b692443f4466d13fc60d5d01a99c22d70/lib/datadog/appsec/waf.rb#L284-L286
- https://github.com/DataDog/libddwaf-rb/pull/33/files#diff-22056d421bdde5b2bdeef273ab884c30efdb1fcc791ac840e8670ff68d675582R491-R493
- https://github.com/DataDog/dd-trace-rb/blob/f7757d28c9424efec937eecf91bd7d8961f6eb41/lib/datadog/appsec/processor.rb#L85

So that would be 4096 bytes:

https://github.com/DataDog/libddwaf/blob/master/include/ddwaf.h#L26

> but which is possibly larger, up until 4x larger

Correct as of today (and for the foreseeable future), although to be nitpicky theoretically designed to be infinitely extensible (quite trivially up to seven bytes).

> Personally, I just set very high limits on the WAF.

Definitely makes sense, bare 100% safe minimum would be 4x indeed.

I'd argue that [`DDWAF_MAX_STRING_LENGTH`](https://github.com/DataDog/libddwaf/blob/master/include/ddwaf.h#LL26C9-L26C32) should be code point oriented and compute that 4x factor to restore the consistency (so default would be 4096 worst case code points, thus 16394 worst-case bytes).